### ISSUE TITLE
bear causing the instrumented build process to deadlock

### ISSUE THREAD

**Issue by TTimo** [rizsotto/Bear#173]
Hello,

I am trying to use bear to capture a fairly complex build process, and finding that when the build is running under bear it eventually deadlocks and stops progressing, no longer spawning new compile processes. Happens after compiling through 100 or so files, out of a build that has ~3000.

Since bear doesn't support incremental and only writes the database at the end of the process, I'm pretty much DOA there. Any idea what could cause this in bear? Other than changing execution timings slightly etc.

**Comment by rizsotto** [comment]
Hey Tim, I've never tried to run Bear myself on such a big codebase. (I should give it a try in the near future.) But heard from users to run against the linux kernel was successful. So, I guess it's not about the number of files, but how complex is the build process.

You are right, Bear does parse the "execution trace" files only after the build is finished. So, it's during the intercept phase. It uses `LD_PRELOAD` which might cause problem if one of the tool is also using it. Or if the build process is running parallel that can potentially increase the risk to make something wrong. And multithreaded build tools (xcode is one known example) might cause problems, since it was not tested against such scenarios.

If you are able to help to fix the issue, I would ask you to maybe share more about the build environment. (What OS? What compiler(s)? What build tools? Is the project source is open?) And also you can experiment with less parallelism in the execution (if there is any) to narrow down the problem space.

**Comment by TTimo** [comment]
This is happening with Unreal Engine 4, which is using a custom build tool (called UBT). After forcing a single compile process at a time I did confirm that the parallelization was indeed related to the hang.

You could get access to the source yourself, it's "free" provided you agree to a EULA with Epic. Eventually I think it would make sense if UBT would produce the compilation database on it's own. I wonder if using it with bear could be fixed though.

Is there a command line option to get clang to spit the compilation database information for a single file directly? Maybe instead of an LD_PRELOAD approach, doing a custom clang invocation would be better?

**Comment by rizsotto** [comment]
Thanks to come back with your findings. Will take a look on the ubt on the weekend.

Clang does not produce compilation database entries. For non LD_PRELOAD tricks, there is a compiler wrapper trick. (I have another project called [scan-build](https://github.com/rizsotto/scan-build) which implement that. The trade off is that the build tool has to respect `CC, `CXX`` environment variables.

**Comment by TTimo** [comment]
I tracked down the cause for the hangs:

tl;dr since Bear is injecting into processes it generally knows nothing about, it can only use multithread safe APIs. I caught the injected processes stuck on setlocale calls, and later on mbstowcs calls.

The setlocale man page reports that it's MT-Unsafe. I assume UBT uses it also, and that collides eventually. mbstowcs should be safe according to the man page, but on my system, that ends up calling mbsrtowcs, which isn't safe either, so I yanked it as well.

I'm having the full build going again now, it's well past the point where it would normally start locking up and it's going just fine.

Not sure how to fix all this the right way. I think libear.so should write the data 'raw' and avoid touching all these functions, move more of the processing logic to the other side..

**Comment by rizsotto** [comment]
Thanks Tim for the bug hunt! Interesting findings, since all `setlocale` and `mbstowcs` calls are guarded with `pthread_mutex_lock` and `pthread_mutex_unlock`... You might be right, one of the earlier version was not care the character encoding at the `libear` module. But then I changed the file format to JSON, which does not allow unicode characters written without escaping it properly... Will think about to change it back.

**Comment by TTimo** [comment]
The mutex will not protect from the process you are injecting into making it's own calls to `setlocale` and `mbstowcs` directly.

You could intercept `setlocale` and `mbstowcs` the same way you are intercepting the spawn functions, and try to apply the mutex there.

Personally I would seek to simplify the libear.so side and make it as safe and re-entrant as possible, offload any locale and encoding processing until after the interception phase.

**Comment by rizsotto** [comment]
Thanks Tim. You are right, if the process calls `setlocale` itself parallel to the one which is in `libear` that is not protected by the mutex. I would not hijack those function calls, simply because I want `libear` hijack OS related calls only, but not C library calls.

I agree, I will simplify the `libear` module as you recommended. It will take some time, but your point is taken. Thanks for help me with this!

**Comment by rizsotto** [comment]
Hey Tim, I just made a new version 2.3.7 which suppose to have a fix about the `setlocale` problem. I'm calling `newlocale` and `uselocale` instead of `setlocale`, which is MT-safe. Could you take a look at it and verify that fix the problem for you? Thanks in advance.

**Comment by TTimo** [comment]
I will check, but since `mbstowcs` is still used I expect it'll still lockup there.

**Comment by TTimo** [comment]
Latest is working great. No lockups anymore, great stuff!

**Comment by TTimo** [comment]
I've done more testing, this problem is fixed for me. Great stuff!

**Comment by rizsotto** [comment]
Thanks Tim for your help! I'm really glad that you have found this bug.
